In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
msroom = pd.read_csv('G:\My Study\Springboard\Projects\Capstone1\Data\mushrooms.csv')
msroom.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color   ...   stalk-surface-below-ring  \
0            c         n          k   ...                          s   
1            c         b          k   ...                          s   
2            c         b          n   ...                          s   
3            c         n          n   ...                          s   
4            w         b          k   ...                          s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [2]:
#Calculate the correlation between every variable and 'class'
from scipy.stats import chi2_contingency
chi = {}
for column in msroom.columns[1:]:
    left = pd.DataFrame(msroom[msroom['class']=='p'][column].value_counts())
    right = pd.DataFrame(msroom[msroom['class']=='e'][column].value_counts())
    joined = left.join(right,lsuffix='_p',rsuffix='_e',how='outer')
    joined[np.isnan(joined)] = 0
    #Using the chi2_contingency() to calculate the chi-square value and p value
    chi[column] = chi2_contingency(joined)[0:2]
chi

{'bruises': (2041.4156474619554, 0.0),
 'cap-color': (387.59776897722986, 6.0558145983365735e-78),
 'cap-shape': (489.91995361895573, 1.196456568593578e-103),
 'cap-surface': (315.04283120803768, 5.5184270386491432e-68),
 'gill-attachment': (133.98618128656679, 5.5017074118610094e-31),
 'gill-color': (3765.7140862414803, 0.0),
 'gill-size': (2366.8342569059605, 0.0),
 'gill-spacing': (984.14333301447391, 5.0229776137324786e-216),
 'habitat': (1573.7772608252621, 0.0),
 'odor': (7659.7267401653389, 0.0),
 'population': (1929.7408909028088, 0.0),
 'ring-number': (374.73683082671153, 4.2357576417230603e-82),
 'ring-type': (2956.6192780575316, 0.0),
 'spore-print-color': (4602.0331700846045, 0.0),
 'stalk-color-above-ring': (2237.8984964488182, 0.0),
 'stalk-color-below-ring': (2152.3908905982812, 0.0),
 'stalk-root': (1344.4405265497817, 7.7020479049435133e-290),
 'stalk-shape': (84.142038265487187, 4.6047462121551919e-20),
 'stalk-surface-above-ring': (2808.2862874186076, 0.0),
 'stalk-s

It can be seen from the result, the variable 'odor' and 'spore-print-color' have the largest chi square value(7659.7 and 4602.0), meaning that these two variables have the largest correlation between edibility. And correlation between the variable 'veil-type' and edibility('class') is 0, this variable can be ignored when try to decide a mushroom's edibility.